istioctl/cmd/istioctl/dashboard.go

```diff
@@ -0,0 +1,501 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"runtime"
+
+	"github.com/spf13/cobra"
+
+	"istio.io/istio/istioctl/pkg/kubernetes"
+
+	// import all known client auth plugins
+	v1 "k8s.io/api/core/v1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	dashboardCmd = &cobra.Command{
+		Use:     "dashboard",
+		Aliases: []string{"dash", "d"},
+		Short:   "Access to Istio web UIs",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// cmd.Println(cmd.UsageString())
+			cmd.HelpFunc()(cmd, args)
+			if len(args) != 0 {
+				return fmt.Errorf("unknown dashboard %q", args[0])
+			}
+
+			return nil
+		},
+	}
+
+	controlZport = 0
+)
+
+func init() {
+	dashboardCmd.AddCommand(kialiDashCmd())
+	dashboardCmd.AddCommand(promDashCmd())
+	dashboardCmd.AddCommand(grafanaDashCmd())
+	dashboardCmd.AddCommand(jaegerDashCmd())
+	dashboardCmd.AddCommand(zipkinDashCmd())
+
+	dashboardCmd.AddCommand(envoyDashCmd())
+	controlz := controlZDashCmd()
+	controlz.PersistentFlags().IntVar(&controlZport, "ctrlz_port", 9876, "ControlZ port")
+	dashboardCmd.AddCommand(controlz)
+	experimentalCmd.AddCommand(dashboardCmd)
+}
+
+// port-forward to Istio System Prometheus; open browser
+func promDashCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "prometheus",
+		Short:   "Open Prometheus web UI",
+		Long:    `Open Istio's Prometheus dashboard`,
+		Example: `istioctl experimental dashboard prometheus`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := kubernetes.NewClient(kubeconfig, configContext)
+			if err != nil {
+				return fmt.Errorf("failed to create k8s client: %v", err)
+			}
+
+			pl, err := prometheusPods(client)
+			if err != nil {
+				return fmt.Errorf("not able to locate Prometheus pod: %v", err)
+			}
+
+			if len(pl.Items) < 1 {
+				return errors.New("no Prometheus pods found")
+			}
+
+			port, err := availablePort()
+			if err != nil {
+				return fmt.Errorf("not able to find available local port: %v", err)
+			}
+			log.Debugf("Using local port: %d", port)
+
+			// only use the first pod in the list
+			promPod := pl.Items[0]
+			fw, readyCh, err := buildPortForwarder(client, client.Config, promPod.Name, istioNamespace, port, 9090)
+			if err != nil {
+				return fmt.Errorf("could not build port forwarder for Prometheus: %v", err)
+			}
+
+			errCh := make(chan error)
+			go func() {
+				errCh <- fw.ForwardPorts()
+			}()
+
+			select {
+			case err := <-errCh:
+				return fmt.Errorf("failure running port forward process: %v", err)
+			case <-readyCh:
+				log.Debugf("port-forward to Prometheus pod ready")
+				defer fw.Close()
+
+				openBrowser(fmt.Sprintf("http://localhost:%d", port), cmd.OutOrStdout())
+
+				// Block forever
+				<-(chan int)(nil)
```

Comment:
I don't think this go routine is tidied up?

Reply:
I am unclear what you are suggesting.  This go routine never terminates.  Please advise.

Reply:
Won't it leak the goroutine? There's no way to terminate this routine, right?

Reply:
I made further improvements to the code hoping to get your positive review.  At the same time you `/approve`d it!  Take another look, let me know.